### PR TITLE
Imviz simple aperture photometry plugin (no photutils)

### LINKED ISSUE #531: JDAT-1400: imexamine-style photometry
Compute photometry for stars selected in image viewer.

https://jira.stsci.edu/browse/JDAT-1400

### PR DESCRIPTION
**Pull request by pllim**
### Description

This pull request is to add a plugin to perform simple aperture photometry in Imviz.

This supersedes #930 . This is a simpler implementation that only works on default viewer and does not rely on `photutils`.

TODO:

- [x] Actually implement the plugin.
- [x] Fix `id` logic so it increments if appending successfully to `QTable`.
- [x] Add change log.
- [x] Polish up POC notebook for demo.
- [x] Use `mode='center'` for normal stats.
- [x] ~Handle pixel area correctly in aperture sum unit.~ New user input field for pixel scale in arcsec squared to be converted to steradians to get rid of it in JWST data unit.
- [x] New user input field for flux to counts conversion factor.
- [x] Update PR screenshots.
- [x] Write tests. [⭐](https://gist.github.com/larrybradley/676370b2637b78e594ad5ed8b27dea86) (Also make sure it works with circle, ellipse, and rectangle. See "Gaussian Smooth" tests for example on how to test plugin.) "devdeps" job failure is unrelated, see #949.
- [x] Update Imviz user-facing doc. ~(Hold off for now until we have finalized the code and UI.)~ See https://jdaviz--938.org.readthedocs.build/en/938/imviz/plugins.html
- ???
- Profit!

(Screenshots a bit outdated but you get the idea.)

<table>
<tr>
<td>

</td>
<td>

</td>
<td>

</td>
</tr>
</table>

<table>
<tr>
<td>

</td>
<td>

</td>
</tr>
</table>

Pre-requisite for #531 .

[🐱](https://jira.stsci.edu/browse/JDAT-1848)

Fixes #531 

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [x] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Did the CI pass? If not, are the failures related?
- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [x] Is a milestone set? Milestone is only currently required for PRs related to Imviz MVP.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

**Review by PatrickOgle — CHANGES_REQUESTED** [review]
Nice!  One request--Please add the option to display the results in counts.  MJy/sr is also not the correct unit.  Need to multiply by the aperture area in sr to get the result in Jy.

**Comment by pllim** [comment]
> How about a pixel scale input field (in arcsec)?

I think I need input from @larrybradley here as `photutils` expert.  `aperture_photometry()` currently does not attempt to multiple output unit by pixel area. Does that mean `aperture_photometry()` output unit is also inaccurate and needs to be fixed? Am I missing something here?

```
<QTable length=2>
  id  xcenter ycenter    aperture_sum
        pix     pix        MJy / sr
int64 float64 float64      float64
----- ------- ------- ------------------
    1    30.0    30.0 28.274333882308134
    2    40.0    40.0 28.274333882308134
```

Also, if we create yet another user input field for pixel scale, is it always applicable for any number of image units, telescope, detector, etc? If someone does not know or care and just want the aperture sum, how should this plugin behave?

**Comment by pllim** [comment]
@PatrickOgle , the value is from image header.

**Comment by PatrickOgle** [comment]
There appears to be a problem with the x, y center values for circular aperture photometry.  The Subset 1 coordinates are (x=1992.08203125, y=1037.489013671875) according to imviz.get_interactive_regions().  The photometry plugin outputs (1991.7647705078125, 1037.489013671875).  In addition, the mouseover pixel x value is way off: (1969.5, 1037.5)

**Comment by PatrickOgle** [comment]
If I leave the Flux scaling box empty, I get the following error when I calculate:
"Aperture photometry failed: ValueError("could not convert string to float:")

**Comment by pllim** [comment]
Responding to (1) from https://github.com/spacetelescope/jdaviz/pull/938#issuecomment-965269285:

* I cannot reproduce the inconsistency between X values reported between `imviz.get_interactive_regions()` and `imviz.get_aperture_photometry_results()`, the latter reflects the results that are also displayed in the plugin. In fact, I have the opposite problem, as reported at glue-viz/glue-astronomy#52 for `imviz.get_interactive_regions()`.
* The coordinates mouseover display being off is a known issue and unrelated to this plugin, see #945 .

**Comment by pllim** [comment]
Following up at https://github.com/spacetelescope/jdaviz/pull/938#issuecomment-965577950 , I think Patrick found a new bug too. See #964

**Comment by pllim** [comment]
Responding to (2) from https://github.com/spacetelescope/jdaviz/pull/938#issuecomment-965269285:

* To have coordinates mouseover display show different formatting for RA and Dec is out of scope here. In fact, in the [Imviz wireframe](https://www.stsci.edu/~jkotler/Imviz/#id=null&p=imviz_ui&g=1), this is not a thing. What I can do here is to change the `SkyCoord` formatting *in the plugin* to match mouseover formatting, but I think you mentioned a preference for decimal, which is not that format, so I am not sure how to proceed here.

**Comment by pllim** [comment]
Responding to comment from 2021-11-10 tag-up about tests: I did include tests as part of this PR. However, in the first pass, given the lack of "real truth" for science data, I basically use the values from the plugin as "truth," which prevents future regression but does not check that this initial implementation is scientifically viable. But thanks to your manual check using DS9, I am now more confident that these tests are correct.

**Comment by pllim** [comment]
Responding to comment from https://github.com/spacetelescope/jdaviz/pull/938#issuecomment-965402797:

* As a dev who wrote it, I see nothing wrong with emitting a "could not convert string to float" error message on the "snackbar" because the user is trying to convert empty string to float when not entering anything in the flux scaling text box (the same behavior apply for other similarly implemented text boxes in the plugin). I *could* make it silently fallback to some default value, but I am afraid that is going to be confusing to the user. As a PO, perhaps you can tell me what is the desired behavior that you want when someone simply deletes a value from the text box. Thanks!

**Comment by PatrickOgle** [comment]
Remaining issues from my above comments:
1) X coordinates in mouseover are incorrect when there are two linked images (to be addressed elsewhere)
2) Mouseover should show both sexigesimal and decimal coordinates (to be addressed elsewhere)
3) Need comprehensive tests that aim for scientific validation, e.g. photometry of fake data that gives the correct results within tolerance for several different sources and apertures. I would be happy for now with photometry of a single noiseless constant surface brightness fake image within square, circular, and elliptical apertures.  A new ticket should  be opened to create a more comprehensive test suite, including aperture photometry more realistic stellar sources.  This is crucial since scientists might be tempted to publish the quicklook photometry tool output without a critical evaluation.
4) I think the "could not convert string to float" error should be more specific if the value is blank, something like "Missing or invalid string value in Background" or "Counts Conversion Factor" or "Pixel Area" or "Flux Scaling".

**Comment by PatrickOgle** [comment]
New requirement  to consider for this PR:
1) Output Poisson uncertainty for sum and mean if counts conversion factor is present.  There may be other uncertainty terms like read noise that we can ignore.  Ideally the uncertainty would be printed in parentheses next to the sum and mean.

**Comment by PatrickOgle** [comment]
I have yet to test the behavior on changing the various numerical input boxes (background, counts conversion, flux scaling, pixel area)...

**Comment by pllim** [comment]
> I would be happy for now with photometry of a single noiseless constant surface brightness fake image within square, circular, and elliptical apertures.

@eteq , please look at the tests in `jdaviz/configs/imviz/tests/test_simple_aper_phot.py` that was added as part of this PR. Do they satisfy the above requirement? Thanks!

> A new ticket should be opened

For my own record: JDAT-1927

**Comment by pllim** [comment]
> Change units for aperture_sum_counts to counts.

I was trying to make it smart by extracting `meta['photometry']['conversion_megajanskys']` from JWST ASDF, which converts to counts/s. I can think of no sane way of being this smart and have it convert to counts no matter what. So, my plan here is to dumb it down and just assume any number given will result in counts conversion, not counts/s.

**Comment by pllim** [comment]
> AB mag

According to https://www.stsci.edu/hst/instrumentation/acs/data-analysis/zeropoints , for the formula to work, the flux needs to be in `F_NU`. So, does this mean, I will need to attempt to convert the aperture sum to erg/s/cm^2/Hz before calculating this magnitude and refuse to do the calculation if that is not possible? Please advise.

**Comment by pllim** [comment]
> TBD: test changing the pixel area for data with units of []/steradian (does nothing for data in electron units)

You can actually test this using the *concept notebook* that is provided by this PR, instead of the default Imviz example notebook. See `notebooks/concepts/imviz_simple_aper_phot.ipynb` (it is not in `main`).

**Comment by PatrickOgle** [comment]
@pllim. Thanks for looking into this. I don't think we should require or convert the flux units to cgs (or Jy) to calculate the magnitude.  Let's just leave it "mag" rather than AB mag so that it is generic.  The user can then decide what number they need to get AB mag out, if that's what they want.  It is  a simple scaling (or magnitude offset) in any case.

**Comment by PatrickOgle** [comment]
@pllim I agree that the conversion factor should convert from data values to counts (not cts/s), and it is up to the user to decide what that needs to be.

**Comment by pllim** [comment]
> Output Poisson uncertainty for sum and mean if counts conversion factor is present. 

@PatrickOgle , this implies that you also want mean in the unit of counts, which is currently not the case. And it seems weird to only compute mean in counts and not the other stats. Can we just compute Poisson uncertainly for `aperture_sum_counts` and forget about the mean?

**Comment by PatrickOgle** [comment]
I am not suggesting to change the units of mean. However,  it is fine if you just give Poisson uncertainty on counts, and not calculate the uncertainty in the mean (which  is just mean * sqrt(counts)/counts)).   I am OK with leaving off the extra output for uncertainty in the mean, to save space.

**Comment by pllim** [comment]
> which is just mean * sqrt(counts)/counts)

Still, it would be confusing to have `mean` displayed in MJy/sr but then in the parentheses have its Poisson uncertainty in counts, no?

> I am OK with leaving off the extra output for uncertainty in the mean, to save space

Okay. In that case, I think this PR has incorporated all your requested changes. Can you please have another look and approve if it is satisfactory? Thanks!

**Comment by rosteen** [comment]
I pushed a small change needed due to #913, hope you don't mind. Will finish my review shortly...

**Review by rosteen — APPROVED** [review]
The code looks good to me, everything looks sensible and works (as far as I can tell, perhaps @PatrickOgle will provide more rigorous validation of the calculation results). Nice addition.

Edit: I also meant to say that I appreciate the detailed documentation!

**Comment by pllim** [comment]
Thanks for the commit, @rosteen ! I do not mind at all. 😸

**Review by PatrickOgle — APPROVED** [review]
The outputs of the photometry tool match the aperture photometry from DS9, to better than 10%.  The Poisson uncertainty on counts also seems to be correct and properly incorporates the flux/counts conversion factor.   More accurate science validation TBD in a separate PR.

**Comment by pllim** [comment]
Thanks for the reviews!
